### PR TITLE
Release 2024.4.1

### LINKED ISSUE #162: Instrumentation and Responsiveness Improvements

### PR DESCRIPTION
**Pull request by josephbirkner**
* Introduce a performance diagnostics dialog which may be opened via `Ctrl+X`
* Fix occasional hanging updates by using mapget's new `clientId` instead of an HTTP abort.
  * Introduced in this mapget PR: https://github.com/ndsev/mapget/pull/79
* Fix the visualization limit, which was blatantly ignored.

TODO:
- [x] Show number of aggregated tiles
- [x] Allow *any* integer/number tile info field to be viewed as a statistic.
- [x] Update statistics automatically (it should not be necessary to click "update").
- [x] Allow viewing render performance separately per style sheet and respective "hover"/"selection" modes.
- [x] Add a Button to open the dialog.
- [x] Do not generate hover highlights while the map is being dragged around.
- [x] Switch to synchronous geometry for tile borders as well.
- [x] Do not prematurely re-encode tiles when only a single one is rendered.
- [x] Statistics "Sum" column.

Closes #162

### CODE REVIEW

**Comment by josephbirkner** [comment]
Future TODOs (partially in other repos):

* Add NDS.Live SmartLayer and DataLayer size and add network fetching wall-time statistics.
* Ensure that region rule layers are not redundantly stored.
* Implement two-stage loading: most fat attribute layers are only needed for search/inspection, but not for visualization.
* Add some kind of loading progress visualization:
  * % of stage 1 layers received
  * % of stage 2 (add-on) layers received
  * % of finished tile visualizations

**Review by Waguramu — DISMISSED** [review]

File: erdblick_app/app/keyboard.service.ts [hunk 1]
```diff
@@ -40,21 +40,23 @@ export class KeyboardService {
...
-                } else if (this.shortcuts.has(key)) {
-                    event.preventDefault();
-                    this.shortcuts.get(key)?.(event);
-                }
+            // TODO: Ensure that tab and escape, when pressed in a text area,
+            //  result in a tab character/autocomplete cancelation rather than
+            //  focusing another control/closing the enclosing dialog.
+
+            // Let non-ctrl key events or text editing shortcuts do their default things.
+            if (isInput && (!key.includes("Ctrl") || ["ctrl+x", "ctrl+c", "ctrl+v"].includes(key.toLowerCase()))) {
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply erdblick_app/app/keyboard.service.ts:48]: - [ ] Add a blacklist for shortcuts that are disabled when the focus is on the input.

**Comment by MisterGC** [comment]
Some improvements for the `Viewport Statistics Dialog`
- [ ] Use floats only when it is applicable, otherwise int (e.g. number of features)
- [ ] Add a tool tip to each entry so that users clearly know what the numbers indicate
- [ ] Improve the labels - e.g. why do the render times for styles always contain `-normal-`, e.g. `Number of Features` instead of always the technical/tracking keyword
image

**Review by MisterGC — APPROVED** [review]
Speed boost achieved! This PR is a significant step forward. Approved!

Although the improvements of the `Viewport Statistics` would help - I have compared cases which still cause performance problems - the peak sizes of the tiles where like 10x higher than ones in areas/viewports/maps (2 vs 20MB)

File: erdblick_app/app/features.model.ts [hunk 2]
```diff
@@ -50,9 +62,12 @@ export class FeatureTile {
     peek(callback: (layer: TileFeatureLayer) => any) {
         // Deserialize the WASM tileFeatureLayer from the blob.
         let result = uint8ArrayToWasm((bufferToRead: any) => {
+            let startTime = performance.now();
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply erdblick_app/app/features.model.ts:62]: Took note, will check.

**Pull request merged**